src/main/java/com/commercetools/sync/services/impl/BaseService.java

```diff
@@ -78,4 +83,55 @@
+     *
+     * @param resource              The resource to update.
+     * @param updateCommandFunction a {@link BiFunction} used to compute the update command required to update the
+     *                              resource.
+     * @param updateActions         the update actions to execute on the resource.
+     * @return an instance of {@link CompletionStage}&lt;{@code U}&gt; which contains as a result an instance of
+     *         the resource {@link U} after all the update actions have been executed.
+     */
+    @Nonnull
+    CompletionStage<U> updateResource(
+        @Nonnull final U resource,
+        @Nonnull final BiFunction<U, List<? extends UpdateAction<U>>, UpdateCommand<U>> updateCommandFunction,
+        @Nonnull final List<UpdateAction<U>> updateActions) {
+        final List<List<UpdateAction<U>>> actionBatches = batchElements(updateActions, MAXIMUM_ALLOWED_UPDATE_ACTIONS);
+        return updateBatches(CompletableFuture.completedFuture(resource), updateCommandFunction, actionBatches);
+    }
+
+    /**
+     * Given a list of update actions batches represented by a {@link List}&lt;{@link List}&gt; of {@link UpdateAction},
+     * this method recursively executes the update command, computed by {@code updateCommandFunction}, on each batch,
+     * then removes the batch, until there are no more batches.
+     *
+     * @param result                in the first call of this recursive method, this result is normally a completed
+     *                              future containing the resource to update, it is then used withing each recursive
+     *                              batch execution to have the latest resource (with version) once the previous batch
+     *                              has finished execution.
+     * @param updateCommandFunction a {@link BiFunction} used to compute the update command required to update the
+     *                              resource.
+     * @param batches               the batches of update actions to execute.
+     * @return an instance of {@link CompletionStage}&lt;{@code U}&gt; which contains as a result an instance of
+     *         the resource {@link U} after all the update actions in all batches have been executed.
+     */
+    @Nonnull
+    private CompletionStage<U> updateBatches(
+        @Nonnull final CompletionStage<U> result,
+        @Nonnull final BiFunction<U, List<? extends UpdateAction<U>>, UpdateCommand<U>> updateCommandFunction,
+        @Nonnull final List<List<UpdateAction<U>>> batches) {
+        if (batches.isEmpty()) {
+            return result;
+        }
+        final List<UpdateAction<U>> nextBatch = batches.remove(0);
+        final CompletionStage<U> nextBatchUpdateFuture = result.thenCompose(updatedProduct ->
+            syncOptions.getCtpClient().execute(updateCommandFunction.apply(updatedProduct, nextBatch))
+        );
+        return updateBatches(nextBatchUpdateFuture, updateCommandFunction, batches);
+    }
```

Comment:
i see here couple related problems:
1. Why did you decide to use recursion? Does it give you any benefit here?
2. `remove(0)` is quite inefficient operation on random access lists, like `ArrayList`: it rebuilds the whole least every time. Of course, for short lists (like in this case) it is not a bid deal, but anyway try not to use such inconvenient approaches in our code.
3. We have a result, which we wrap to `CompletionStage` and then immediately compose -> unnecessary wrapping. Wrap it to `CompletionStage` only when return the result in `if (batches.isEmpty())`

Reply:
1. I have used recursion not because it gives me any benefit other than it was the most straightforward and cleanest way for me to implement processing of on an array of completion stages that depend the previous element's completion stage :) better suggestions?
2. IMO, it is completely justified in this case with the worst case complexity being O(n) for removal (with the shifting) on the most probably very small sizes of batches. (the update actions are only per one product (comparing one product to another). Also I am not sure, how to be better. I mean it could be O(1) If I remove the last batch first, but then to do that I would need to reverse the list (and that is also linear O(n) ).
3. I really don't see how it's unnecessary. (Or maybe I don't understand this point at all :) ) But I the completion stage is there because its needed for the recursive call so that the subsequent calls are only executed after the execution of the earlier completion stages.

Reply:
1. IMHO recursion is always confusing, especially if the data structure is not recursive (like graphs), and usually more error prone.
2. Question is not about complexity, but about code style - removing zero element looks suspicious and usually used in some complex logic
3. Then it is one more point to avoid recursion :)

Reply:
My proposition:
```java 
private CompletionStage<U> updateBatches(
    @Nonnull final CompletionStage<U> result,
    @Nonnull final BiFunction<U, List<? extends UpdateAction<U>>, UpdateCommand<U>> updateCommandFunction,
    @Nonnull final List<List<UpdateAction<U>>> batches) {
    if (batches.isEmpty()) {
        return result;
    }

    CompletionStage<U> completionStage = result;

    for (final List<UpdateAction<U>> batch : batches) {
        completionStage = completionStage.thenCompose(updatedProduct ->
            syncOptions.getCtpClient().execute(updateCommandFunction.apply(updatedProduct, batch)));
    }

    return completionStage;
}
```

Reply:
604de5d